src/testing.c

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright (C) 2016 The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file LICENSE for details.
+*/
+
+#include "testing.h"
+
+#include "console.h"
+
+// A flag used to record the return value of the most recent test
+static int recent_test_result = 1;
+
+// A function to time a specific test
+// Returns the amount of time that elapsed during the test
+clock_t time_test(tester test) {
+	clock_t start = clock_read();
+	recent_test_result = (test)();
+	clock_t end = clock_read();
+	return clock_diff(start, end);
+}
+
+// A test runner
+// Returns true if a test was executed in a passing amount of time
+int run_test(struct test_runner *runner) {
+	clock_t actual_time = time_test(runner->test);
+	if(clock_compare(actual_time, runner->runtime_bound) < 0 || !recent_test_result) {
+		// The test failed
+		console_printf("Test %s FAILED with a time of %d.%ds\n", runner->title, actual_time.seconds, actual_time.millis);
+		return 0;
+	} else {
+		console_printf("Test %s passed with a time of %d.%ds\n", runner->title, actual_time.seconds, actual_time.millis);
+		return 1;
+	}
+}
+
+void run_all_tests() {
+	int size = sizeof(tests) / sizeof(tests[0]);
+	console_printf("Running %d tests\n", size);
+	int i;
+	int failed = 0;
+	for(i = 0; i < size; i += 1) {
+		if(!run_test(&(tests[i]))) {
```

Comment:
The rest runner needs to be passed by reference, as when we pass it by value, we get the general protection fault which I believe is a result of trying to copy the struct.